### PR TITLE
New transport with reader writer to support stream.

### PR DESCRIPTION
**Pull request by qiwzhang**

### CODE REVIEW

File: include/transport.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+namespace istio {
+namespace mixer_client {
+
+// A stream write interface implemented by transport layer
+// It will be called by Mixer client.
+template <class RequestType>
+class WriteInterface {
+ public:
+  // Write a request message.
+  virtual void Write(const RequestType&) = 0;
```
Comment (lizan) [inline include/transport.h]: How does caller know if write succeed (or the writer accepts messages)?

File: include/transport.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  virtual void OnClose(const ::google::protobuf::util::Status&) = 0;
+};
+
+// This is the transport interface needed by Mixer client.
+// The callers of the Mixer client need to implement this interface and
+// pass it to the client.
+class TransportInterface {
+ public:
+  // Create a new check stream
+  typedef std::shared_ptr<WriteInterface<::istio::mixer::v1::CheckRequest>>
```
Comment (lizan) [inline include/transport.h]: Not sure `std::shared_ptr` is the right pointer we want to use.

`std::unique_ptr` for Writers, raw pointers for Readers should be enough for this interface.

File: include/transport.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  virtual void OnRead(const ResponseType&) = 0;
+  // On stream close.
+  virtual void OnClose(const ::google::protobuf::util::Status&) = 0;
+};
+
+// This is the transport interface needed by Mixer client.
+// The callers of the Mixer client need to implement this interface and
+// pass it to the client.
+class TransportInterface {
+ public:
```
Comment (lizan) [inline include/transport.h]: virtual destructor

File: include/transport.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  // Half close the write direction.
+  virtual void WriteDone() = 0;
+};
+
+// A stream read interface implemented by Mixer client
+// to receive response data, or stream close status.
+// It will be called by the transport layer.
+template <class ResponseType>
+class ReadInterface {
+ public:
```
Comment (lizan) [inline include/transport.h]: virtual destructor

File: include/transport.h (hunk 1 above)
Comment (qiwzhang) [inline include/transport.h]: Such error should cause stream to be closed.  So reader->OnClose(status) will be called; a single point of close/error handling.

File: include/transport.h (hunk 4 above)
Comment (qiwzhang) [inline include/transport.h]: Done

File: include/transport.h (hunk 3 above)
Comment (qiwzhang) [inline include/transport.h]: Done

File: include/transport.h (hunk 2 above)
Comment (qiwzhang) [inline include/transport.h]: Done.  Thanks it is much easier to use for raw pointer.

File: include/transport.h (hunk 1 above)
Comment (lizan) [inline include/transport.h]: reader->OnClose(status) means stream is full closed (no longer writable or readable), but how will the writer knows the stream is half closed (no longer writable)?

e.g. To support transport doesn't support sending stream request, the writer will be immediately marked as half closed when it gets first request, but Reader->OnClose should be called after OnRead.

**Comment by qiwzhang** [comment]
This is very tricky.   The issue is transport interface is True streaming,
 but the underneath transport layer doesn't supports it, it only supports
unary.   If we want to support this:
The request/response pairing layer;  StreamTransport,  needs to support
multiple streams, one for the current stream, and many half-closed ones.  I
need to re-design that one.

Do we really need to support this case?   For now, with only unary, we
don't use this transport interface.  Once the true streaming is supported,
we will start to use the interface.

**Comment by qiwzhang** [comment]
OK, I re-design-ed the StreamTransport in src/transport_impl.h to support half close.

**Comment by qiwzhang** [comment]
Dont worry about data protection for multi-threads for now.  It will be done in a separate PR.

**Pull request merged**